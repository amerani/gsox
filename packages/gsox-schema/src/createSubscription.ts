import "reflect-metadata";
import { FIELD_SYMBOL, TYPE_SYMBOL } from "./constants";

function createSubscription(T): string {
      const obj = new T();
      const type = Reflect.get(obj, TYPE_SYMBOL);
      let schema = `subscription {\n\t${type} {\n`;
      const metadata = Reflect.getMetadata(FIELD_SYMBOL, obj);
      metadata.forEach((field) => {
            schema += `\t\t${field.name}\n`;
      });
      schema += `\t}\n}`;
      return schema;
}

export {
      createSubscription,
};
