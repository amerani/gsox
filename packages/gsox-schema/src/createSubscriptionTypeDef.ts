import { TYPE_SYMBOL } from "./constants";

function createSubscriptionType(types) {
      const typeNames = types
            .map((type) => new type())
            .map((instance) => Reflect.get(instance, TYPE_SYMBOL));

      let result = `type Subscription {\n`;
      typeNames.forEach((typeName) => {
            result = result.concat(`${typeName}:${typeName}\n`);
      });
      result = result.concat("}");
      return result;
}

export { createSubscriptionType };
