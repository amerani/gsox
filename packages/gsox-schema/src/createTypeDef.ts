import "reflect-metadata";
import { FIELD_SYMBOL, TYPE_SYMBOL } from "@gsox/core";

function createTypeDef(T): string {
      const obj = new T();
      const type = Reflect.get(obj, TYPE_SYMBOL);
      let schema = `type ${type} {\n`;
      const metadata = Reflect.getMetadata(FIELD_SYMBOL, obj);
      metadata.forEach((field) => {
            schema += `\t${field.name}: ${field.type}\n`;
      });
      schema += `}`;
      return schema;
}

function createTypeDefs(types): string {
      return types.reduce(
            (acc, cur) => {
                  const typeStr = createTypeDef(cur);
                  return acc.concat([typeStr, "\n"]);
            },
            "");
}

export {
      createTypeDef,
      createTypeDefs,
};
