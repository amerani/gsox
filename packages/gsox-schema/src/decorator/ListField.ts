import "reflect-metadata";
import { FIELD_SYMBOL, TYPE_SYMBOL } from "@gsox/core";

function ListField<T extends {new():{}}>(type: T, name?) {
    return function(obj, prop) {
        console.log("fields run first bro")
        const id = prop;
        name = name || prop;
        let typeName, children = []

        const listType = type as T;
        const instance = new listType();
        console.log('type decorator must have set type name???')
        if(instance) {
            typeName = `[${Reflect.get(instance, TYPE_SYMBOL) || listType.name}]`;
            children = children.concat(Reflect.getMetadata(FIELD_SYMBOL, instance));
        }

        const entry = { id, typeId: listType.name, name, type: typeName, children };

        if (Reflect.hasMetadata(FIELD_SYMBOL, obj)) {
            const cur = Reflect.getMetadata(FIELD_SYMBOL, obj);
            Reflect.defineMetadata(FIELD_SYMBOL, [entry, ...cur], obj);
        } else {
            Reflect.defineMetadata(FIELD_SYMBOL, [entry], obj);
        }
    };
}

export { ListField };