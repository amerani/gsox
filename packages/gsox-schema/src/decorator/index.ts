export * from "./ListField";
export * from "./Type";