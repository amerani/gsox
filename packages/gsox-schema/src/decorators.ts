import { FIELD_SYMBOL, TYPE_SYMBOL } from "@gsox/core";
import "reflect-metadata";

function Type(value?) {
      return function classDecorator<T extends {new(...args: any[]): {}}>(constructor: T) {
            return class extends constructor {
                  public [TYPE_SYMBOL] = value || constructor.name;
            };
        };
}

function Field(type?, name?) {
      return function(obj, prop) {
            const id = prop;
            name = name || prop;
            type = type || Reflect.getMetadata("design:type", obj, prop).name;
            const entry = {id, name, type};
            if (Reflect.hasMetadata(FIELD_SYMBOL, obj)) {
                  const cur = Reflect.getMetadata(FIELD_SYMBOL, obj);
                  Reflect.defineMetadata(FIELD_SYMBOL, [entry, ...cur], obj);
            } else {
                  Reflect.defineMetadata(FIELD_SYMBOL, [entry], obj);
            }
      };
}

export {
      Type,
      Field,
};
