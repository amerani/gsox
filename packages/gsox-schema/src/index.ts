export * from "./decorators";
export * from "./typeDef";
export * from "./subscription";
export * from "./typeSub";