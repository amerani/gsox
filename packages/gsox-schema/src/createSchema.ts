import gql from "graphql-tag";
import "reflect-metadata";
import { createSubscriptionType } from "./createSubscriptionTypeDef";
import { createTypeDefs } from "./createTypeDef";

const rootSchema = `
      type Query {
            hello: String
      }
      type schema {
            subscription: Subscription
            query: Query
      }
      type Ping {
            id: Int
      }
      type Subscription {
            Ping: Ping
      }
`;
function createTypeDefString(types) {
      return `
            ${createSubscriptionType(types)} \n
            ${createTypeDefs(types)}
      `;
}

function createSchemaString(types) {
      return `
            ${rootSchema} \n
            ${createTypeDefString(types)}
      `;
}

function createSchema(types) {
      return gql(createSchemaString(types));
}

export {
      createSchemaString,
      createSchema,
};
