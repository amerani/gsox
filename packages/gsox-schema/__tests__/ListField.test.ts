import { Type, ListField } from '../src/decorator';
import { Field } from '../src/decorators';
import { createSchemaString } from '../src/createSchema';

@Type()
class AlertType {
    @Field()
    name:string;
}

@Type()
class Message {
    @Field()
    text: string;
}

@Type('OnCallAlert')
class Alert {
    @ListField(Alert)
    previous: Alert[];
    @Field('Int')
    id:number;
    @Field(AlertType)
    type: AlertType;
    @ListField(Message)
    messages: Message[]
}

test('should create schema', () => {
    const schema = createSchemaString([Alert, AlertType, Message]);
    expect(schema).toMatchSnapshot();
})