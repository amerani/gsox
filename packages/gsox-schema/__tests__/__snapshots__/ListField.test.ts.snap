// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create schema 1`] = `
"
            
      type Query {
            hello: String
      }
      type schema {
            subscription: Subscription
            query: Query
      }
      type Ping {
            id: Int
      }
      type Subscription {
            Ping: Ping
      }
 

            
            extend type Subscription {
OnCallAlert:OnCallAlert
AlertType:AlertType
Message:Message
} 

            type OnCallAlert {
	messages: [Message]
	type: AlertType
	id: Int
	previous: [OnCallAlert]
},
type AlertType {
	name: String
},
type Message {
	text: String
},

      
      "
`;
